src/Data/JoinSemilattice/Class/Eq.hs

```diff
@@ -12,42 +13,39 @@ License     : MIT
 module Data.JoinSemilattice.Class.Eq where
 
 import Control.Applicative (liftA2)
-import Data.Hashable (Hashable)
 import Data.JoinSemilattice.Class.Boolean (BooleanR (..))
-import Data.JoinSemilattice.Class.Merge (Merge)
 import Data.JoinSemilattice.Defined (Defined (..))
-import Data.JoinSemilattice.Intersect (Intersect (..))
+import Data.JoinSemilattice.Intersect (Intersect (..), Intersectable)
 import qualified Data.JoinSemilattice.Intersect as Intersect
-import Data.Kind (Type)
+import Data.Kind (Constraint, Type)
 
 -- | Equality between two variables as a relationship between them and their
 -- result. The hope here is that, if we learn the output before the inputs, we
 -- can often "work backwards" to learn something about them. If we know the
 -- result is exactly /true/, for example, we can effectively then
 -- 'Control.Monad.Cell.Class.unify' the two input cells, as we know that their
 -- values will always be the same.
-class (BooleanR b, Merge x) => EqR (x :: Type) (b :: Type) | x -> b where
```

Comment:
As remarked above, add these back to keep other signature shorter.

Reply:
So, trying to do that, I remembered why I removed them: We don't really have `x` available to constrain anymore with the new approach.

Perhaps the right thing to do would be to also adapt `Merge` and `BooleanR`, along the lines of:

```
class Merge f c | f -> c where
  (<<-) :: c x => f x -> f x -> Result (f x)
```

I'm having issues as remarked below aligning the various constraints for the different classes though.

Reply:
I think I'd rather just add `Merge x => ...` to the type signature within the class methods - `Merge` works on `*` things for reasons like `()`, `All`, `Any`, etc

Reply:
I've done this for `EqR` in https://github.com/i-am-tom/holmes/pull/11/commits/b594f439575f7721c85ae54f3bf8289ab6ef69ab, is that what you had in mind? It doesn't seem to help with the type signatures.